File: backend/src/grouter/routetable.c
/*
 * routetable.c (Route table) routines
 * AUTHOR: Revised by Muthucumaru Maheswaran
 * DATE:   Revised on December 23, 2004
 */

#include "routetable.h"
#include "gnet.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <slack/err.h>



/*-------------------------------------------------------------------------
 *                   R O U T E  T A B L E  F U N C T I O N S
 *-------------------------------------------------------------------------*/

/*
 * These route table functions are not very efficient.. I guess because
 * of the small topologies that are targeted here... efficiency is not
 * a major concern...
 */


int rtbl_replace_indx;         // indicate which entry in routing table should be replaced


/*
 * Find an interface corresponding to an IP address
 * Result stored in pbNhop and ppsInterfaceRet
 * Returns NO_ERROR if match found, ERROR if no match found
 */
int findRouteEntry(route_entry_t route_tbl[], uchar *ip_addr, uchar *nhop, int *ixface)
{
	int icount;
	uchar null_ip_addr[] = {0, 0, 0, 0};
	char tmpbuf[MAX_TMPBUF_LEN];
	/* Longest-prefix match in a SINGLE pass — track the best (widest mask) match inline.
	 * The old version collected matches into a fixed mindex[4] and could overflow when more
	 * than 4 routes matched (>4 overlapping subnets); this cannot overflow. */
	int k = -1;
	int nmlen = -1;

	for (icount = 0; icount < MAX_ROUTES; icount++)
	{
		if (route_tbl[icount].is_empty == TRUE) continue;
		verbose(2, "[findRouteEntry]:: Testing the route entry.. g %s at RT[%d], net %s netmask %s nexthop %s int %d",
			       IP2Dot(tmpbuf, ip_addr), icount, IP2Dot(tmpbuf+15, route_tbl[icount].network),
			       IP2Dot(tmpbuf+30, route_tbl[icount].netmask), IP2Dot(tmpbuf+45, route_tbl[icount].nexthop),
			       route_tbl[icount].interface);
		if (compareIPUsingMask(ip_addr, route_tbl[icount].network, route_tbl[icount].netmask) == 0)
		{
			int l = netMaskLen(route_tbl[icount].netmask);
			verbose(2, "[findRouteEntry]:: Found a route for %s at RT[%d], net %s netmask %s nexthop %s int %d",
			       IP2Dot(tmpbuf, ip_addr), icount, IP2Dot(tmpbuf+15, route_tbl[icount].network),
			       IP2Dot(tmpbuf+30, route_tbl[icount].netmask), IP2Dot(tmpbuf+45, route_tbl[icount].nexthop),
			       route_tbl[icount].interface);
			if (l > nmlen) { nmlen = l; k = icount; }
		}
	}

	if (k >= 0)
	{
		if (COMPARE_IP(route_tbl[k].nexthop, null_ip_addr) == 0)
			COPY_IP(nhop, ip_addr);
		else
			COPY_IP(nhop, route_tbl[k].nexthop);

		*ixface = route_tbl[k].interface;

		return EXIT_SUCCESS;
	}
	else
	{
		verbose(2, "[findRouteEntry]:: No match for %s in route table", IP2Dot(tmpbuf, ip_addr));
		return EXIT_FAILURE;
	}
}


/*
 * Add a route entry to the table, if entry found update, else fill in an empty one.
 * If no empty entry, overwrite a used one (round-robin via rtbl_replace_indx) — but
 * NEVER a CONNECTED entry: evicting a router's own interface subnet silently destroys
 * its connectivity (a hazard when a buggy control-plane protocol floods the table).
 * If every slot is connected, the add is refused loudly instead.
 */
void addRouteEntryTagged(route_entry_t route_tbl[], uchar* nwork, uchar* nmask, uchar* nhop,
                         int interface, uchar origin)
{
	int i;
	int ifree = -1;
	char tmpbuf[MAX_TMPBUF_LEN];

	// First check if the entry is already in the table, if it is, update it
	for (i = 0; i < MAX_ROUTES; i++)
	{
		if (route_tbl[i].is_empty == TRUE)
		{
			if (ifree < 0) ifree = i;

		} else if ((COMPARE_IP(nwork, route_tbl[i].network)) == 0)
		{
			if ((COMPARE_IP(nmask, route_tbl[i].netmask)) == 0)
			{
				// match: last writer wins, and the origin tag records who
				COPY_IP(route_tbl[i].nexthop, nhop);
				route_tbl[i].interface = interface;
				route_tbl[i].origin = origin;

				verbose(2, "[addRouteEntry]:: updated route table entry #%d", i);
				return;
			}
		}
	}

	if (ifree < 0)
	{
		// table full: round-robin eviction, skipping CONNECTED entries
		for (i = 0; i < MAX_ROUTES; i++)
		{
			int cand = (rtbl_replace_indx + i) % MAX_ROUTES;
			if (route_tbl[cand].origin != ROUTE_ORIGIN_CONNECTED)
			{
				ifree = cand;
				rtbl_replace_indx = (cand + 1) % MAX_ROUTES;
				break;
			}
		}
		if (ifree < 0)
		{
			error("[addRouteEntry]:: route table full of connected routes; cannot add %s",
			      IP2Dot(tmpbuf, nwork));
			return;
		}
	}

	COPY_IP(route_tbl[ifree].network, nwork);
	COPY_IP(route_tbl[ifree].netmask, nmask);
	COPY_IP(route_tbl[ifree].nexthop, nhop);
	route_tbl[ifree].interface = interface;
	route_tbl[ifree].origin = origin;
	route_tbl[ifree].is_empty = FALSE;

	verbose(2, "[addRouteEntry]:: wrote route entry #%d (origin %c)", ifree, origin);
	return;
}


/*
 * Untagged add (existing callers): the origin is derived — a zero nexthop is a
 * CONNECTED (interface) route, anything else is STATIC (CLI/boot `route add`).
 * Control-plane protocols use addRouteEntryTagged(..., ROUTE_ORIGIN_DYNAMIC).
 */
void addRouteEntry(route_entry_t route_tbl[], uchar* nwork, uchar* nmask, uchar* nhop, int interface)
{
	uchar null_ip_addr[] = {0, 0, 0, 0};
	uchar origin = (COMPARE_IP(nhop, null_ip_addr) == 0) ?
	               ROUTE_ORIGIN_CONNECTED : ROUTE_ORIGIN_STATIC;
	addRouteEntryTagged(route_tbl, nwork, nmask, nhop, interface, origin);
}



/*
 * delete route table entry by argument index i
 */
void deleteRouteEntryByIndex(route_entry_t route_tbl[], int i)
{
	route_tbl[i].is_empty = TRUE;
	verbose(2, "[deleteRouteEntryByIndex]:: route entry #%d deleted", i);
	return;
}


/*
 * delete route table entries related to
 * interface specified by argument indx
 */
void deleteRouteEntryByInterface(route_entry_t route_tbl[], int interface)
{
	int i;

	for (i = 0; i < MAX_ROUTES; i++)
		if ((route_tbl[i].is_empty == FALSE) &&
		    (route_tbl[i].interface == interface))
			deleteRouteEntryByIndex(route_tbl, i);

	verbose(2, "[deleteRouteEntryByInterface]:: table cleared of references to interface: %d", interface);
	return;
}


/*
 * initialize the route table to be empty
 */
void RouteTableInit(route_entry_t route_tbl[])
{
	int i;

	rtbl_replace_indx = 0;

	for(i = 0; i < MAX_ROUTES; i++)
	{
		route_tbl[i].is_empty = TRUE;
		route_tbl[i].origin = ROUTE_ORIGIN_STATIC;
	}
	verbose(2, "[initRouteTable]:: table initialized");

	return;
}


/*
 * print the route table
 */
void printRouteTable(route_entry_t route_tbl[])
{
	int i, rcount = 0;
	char tmpbuf[MAX_TMPBUF_LEN];
	interface_t *iface;

	printf("\n=================================================================\n");
	printf("      R O U T E  T A B L E \n");
	printf("-----------------------------------------------------------------\n");
	printf("Index\tNetwork\t\tNetmask\t\tNexthop\t\tInterface\tOrigin \n");

	for (i = 0; i < MAX_ROUTES; i++)
		if (route_tbl[i].is_empty != TRUE)
		{
			iface = findInterface(route_tbl[i].interface);
			printf("[%d]\t%s\t%s\t%s\t\t%s\t%c\n", i, IP2Dot(tmpbuf, route_tbl[i].network),
			       IP2Dot((tmpbuf+20), route_tbl[i].netmask), IP2Dot((tmpbuf+40), route_tbl[i].nexthop),
			       iface->device_name, route_tbl[i].origin ? route_tbl[i].origin : '?');
			rcount++;
		}
	printf("-----------------------------------------------------------------\n");
	printf("      %d number of routes found. \n", rcount);
	return;
}

File: backend/src/grouter/routetable.zig
//! routetable.zig — the gRouter route table, ported from C to Zig (Z3).
//!
//! Exports the same C ABI as the original routetable.c, against the same
//! `route_entry_t` layout (verified: bool=1B, interface at offset 16, sizeof 20), so
//! callers (ip.c, gr_state.c, cli.c) link against it unchanged. The pretty-printer
//! `printRouteTable` stays in C (it reaches into interface_t); everything else — the
//! data-plane logic — is here, memory-safe (ReleaseSafe: bounds checks on every access).
const std = @import("std");

const MAX_ROUTES: usize = 20;
const MAX_ROUTES_C: c_int = 20;
const TRUE: u8 = 1;
const FALSE: u8 = 0;
const EXIT_SUCCESS: c_int = 0;
const EXIT_FAILURE: c_int = 1;

// matches route_entry_t in routetable.h exactly (size 20).
const RouteEntry = extern struct {
    is_empty: u8,
    network: [4]u8,
    netmask: [4]u8,
    nexthop: [4]u8,
    interface: c_int,
};

// helpers that stay in C (non-variadic, pure).
extern fn compareIPUsingMask(ip: [*c]const u8, network: [*c]const u8, netmask: [*c]const u8) c_int;
extern fn netMaskLen(netmask: [*c]const u8) c_int;

// round-robin overwrite cursor (was a C global of the same name).
export var rtbl_replace_indx: c_int = 0;

inline fn ipEq(a: [*c]const u8, b: [*c]const u8) bool {
    return a[0] == b[0] and a[1] == b[1] and a[2] == b[2] and a[3] == b[3];
}
inline fn ipCopy(dst: [*c]u8, src: [*c]const u8) void {
    dst[0] = src[0];
    dst[1] = src[1];
    dst[2] = src[2];
    dst[3] = src[3];
}

export fn RouteTableInit(route_tbl: [*c]RouteEntry) void {
    rtbl_replace_indx = 0;
    var i: usize = 0;
    while (i < MAX_ROUTES) : (i += 1) route_tbl[i].is_empty = TRUE;
}

// longest-prefix match. (Cleaner than the C original, which tracked at most 4 matches
// in a fixed array — a latent overflow with >4 matching routes.)
export fn findRouteEntry(route_tbl: [*c]RouteEntry, ip_addr: [*c]u8, nhop: [*c]u8, ixface: *c_int) c_int {
    var best: isize = -1;
    var best_len: c_int = -1;
    var i: usize = 0;
    while (i < MAX_ROUTES) : (i += 1) {
        if (route_tbl[i].is_empty != FALSE) continue;
        if (compareIPUsingMask(ip_addr, &route_tbl[i].network, &route_tbl[i].netmask) == 0) {
            const l = netMaskLen(&route_tbl[i].netmask);
            if (l > best_len) {
                best_len = l;
                best = @intCast(i);
            }
        }
    }
    if (best < 0) return EXIT_FAILURE;
    const k: usize = @intCast(best);
    const zero = [_]u8{ 0, 0, 0, 0 };
    if (ipEq(&route_tbl[k].nexthop, &zero)) ipCopy(nhop, ip_addr) else ipCopy(nhop, &route_tbl[k].nexthop);
    ixface.* = route_tbl[k].interface;
    return EXIT_SUCCESS;
}

export fn addRouteEntry(route_tbl: [*c]RouteEntry, nwork: [*c]u8, nmask: [*c]u8, nhop: [*c]u8, interface: c_int) void {
    var ifree: isize = -1;
    var i: usize = 0;
    while (i < MAX_ROUTES) : (i += 1) {
        if (route_tbl[i].is_empty != FALSE) {
            if (ifree < 0) ifree = @intCast(i);
        } else if (ipEq(nwork, &route_tbl[i].network) and ipEq(nmask, &route_tbl[i].netmask)) {
            ipCopy(&route_tbl[i].nexthop, nhop); // update existing
            route_tbl[i].interface = interface;
            return;
        }
    }
    var slot: usize = undefined;
    if (ifree < 0) {
        slot = @intCast(rtbl_replace_indx);
        rtbl_replace_indx = @rem(rtbl_replace_indx + 1, MAX_ROUTES_C);
    } else slot = @intCast(ifree);
    ipCopy(&route_tbl[slot].network, nwork);
    ipCopy(&route_tbl[slot].netmask, nmask);
    ipCopy(&route_tbl[slot].nexthop, nhop);
    route_tbl[slot].interface = interface;
    route_tbl[slot].is_empty = FALSE;
}

export fn deleteRouteEntryByIndex(route_tbl: [*c]RouteEntry, i: c_int) void {
    route_tbl[@intCast(i)].is_empty = TRUE;
}

export fn deleteRouteEntryByInterface(route_tbl: [*c]RouteEntry, interface: c_int) void {
    var i: usize = 0;
    while (i < MAX_ROUTES) : (i += 1) {
        if (route_tbl[i].is_empty == FALSE and route_tbl[i].interface == interface)
            route_tbl[i].is_empty = TRUE;
    }
}

// panic handler: a safety trip aborts cleanly without needing Zig's full runtime.
pub const panic = std.debug.FullPanic(struct {
    fn p(_: []const u8, _: ?usize) noreturn {
        @trap();
    }
}.p);
